lib/Sema/CSStep.cpp

```diff
@@ -362,22 +362,60 @@ StepResult ComponentStep::take(bool prevFailed) {
     return finalize(/*isSuccess=*/false);
   }
 
+#ifdef NDEBUG
+  auto printConstraints = [&](const ConstraintList &constraints) {
+    for (auto &constraint : constraints)
+      constraint.print(getDebugLogger(), &CS.getASTContext().SourceMgr);
+  };
+#endif
+
   // If we don't have any disjunction or type variable choices left, we're done
   // solving. Make sure we don't have any unsolved constraints left over, using
-  // report_fatal_error to make sure we trap in release builds instead of
-  // potentially miscompiling.
+  // report_fatal_error to make sure we trap in debug builds and fail the step
+  // in release builds.
   if (!CS.ActiveConstraints.empty()) {
+#ifndef NDEBUG
     CS.print(llvm::errs());
     llvm::report_fatal_error("Active constraints left over?");
+#else
+    if (CS.isDebugMode()) {
+      getDebugLogger() << "(failed due to remaining active constraints:\n";
+      printConstraints(CS.ActiveConstraints);
+      getDebugLogger() << ")\n";
+    }
+
+    return finalize(/*isSuccess=*/false);
+#endif
   }
+
   if (!CS.solverState->allowsFreeTypeVariables()) {
     if (!CS.InactiveConstraints.empty()) {
+#ifndef NDEBUG
       CS.print(llvm::errs());
       llvm::report_fatal_error("Inactive constraints left over?");
+#else
+      if (CS.isDebugMode()) {
+        getDebugLogger() << "(failed due to remaining inactive constraints:\n";
+        printConstraints(CS.InactiveConstraints);
+        getDebugLogger() << ")\n";
+      }
+
+      return finalize(/*isSuccess=*/false);
+#endif
     }
+
     if (CS.hasFreeTypeVariables()) {
+#ifndef NDEBUG
       CS.print(llvm::errs());
       llvm::report_fatal_error("Free type variables left over?");
+#else
+      if (CS.isDebugMode()) {
+        getDebugLogger() << "(failed due to remaining free type variables)\n";
+        CS.print(getDebugLogger(/*indent=*/false));
+      }
+
+      return finalize(/*isSuccess=*/false);
+#endif
     }
```

Comment:
I think this entire condition is actually dead code, as it's already handled in an above `if` statement, I guess I must have missed that when adding it

Reply:
Ah, I see yeah, let's remove it.

Reply:
Fixed!